### PR TITLE
[terra-avatar] [terra-image] Avatar to allow non square images.

### LINKED ISSUE #2259: [terra-avatar] should allow non-square images
# Feature Request

## Description
Current avatar accepts images, but distorts the presentation unless the image provided is a perfect square. https://github.com/cerner/terra-core/pull/2224#issuecomment-469676287

Need to determine if terra-avatar should allow for non-square images to be handled such that they do not distort the image/photo provided. 

## Additional Context / Screenshots

screen shot 2019-03-05 at 8 53 56 am

### PR DESCRIPTION
**Pull request by pranav300**
### Summary
Resolves #2259

### Additional Details
I have added a prop `fit` in terra-image which would be used to apply `object-fit` styles. Also a polyfill [object-fit-images](https://github.com/bfred-it/object-fit-images#usage) is added to terra-image as object-fit property is not supported in IE. In Avatar and ProfileImage I am setting the imageFit prop as `cover` in Terra-Image tag.

### CODE REVIEW

File: packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx [hunk 1]
```diff
@@ -15,11 +15,11 @@ const ImageFluid = () => (
     </div>
     <p>Image width is larger than the container. Image height is smaller than the container.</p>
     <div id="height-smaller" className={cx('image-conatiner-reduced-width')}>
-      <Image src={placeholder150x150} alt="rounded" variant="rounded" isFluid />
+      <Image src={placeholder150x150} alt="rounded" variant="rounded" imageFit="cover" isFluid />
     </div>
     <p>Image width is smaller than the container. Image height is larger than the container.</p>
     <div id="width-smaller" className={cx('image-conatiner-reduced-height')}>
-      <Image src={placeholder150x150} alt="rounded" variant="rounded" isFluid />
+      <Image src={placeholder150x150} alt="rounded" variant="rounded" imageFit="cover" isFluid />
```
Comment (ShettyAkarsh) [inline packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx]: we could prabably add an test here for imagFit  prop values "contain","scaledown" and "none".

@neilpfeiffer , @ryanthemanuel  your thoughts.

File: packages/terra-image/src/Image.module.scss [hunk 2]
```diff
@@ -36,4 +36,29 @@
   .hidden {
     display: none;
   }
+
```
Comment (nramamurth) [inline packages/terra-image/src/Image.module.scss]: I think it's worth adding inline documentation here explaining why `font-family` is introduced and why it has 'object-fit'.

File: packages/terra-avatar/src/common/AvatarUtils.jsx [hunk 3]
```diff
@@ -71,6 +71,9 @@ const generateImagePlaceholder = (alt, isAriaHidden, variant) => {
  */
 const generateImage = (image, alt, isAriaHidden, variant, handleFallback) => {
   const icon = generateImagePlaceholder(alt, isAriaHidden, variant);
+  if (variant === 'user') {
```
Comment (nramamurth) [inline packages/terra-avatar/src/common/AvatarUtils.jsx]: This could be simplified, I think. 

```
const imageFit = (variant === 'user') ? ImageFitTypes.COVER : undefined;
return <TerraImage className={cx('image')} src={image} placeholder={icon} alt={alt} onError={handleFallback} imageFit={imageFit} />
```

File: packages/terra-image/src/Image.jsx [hunk 4]
```diff
@@ -49,20 +58,27 @@ const propTypes = {
...
   alt: ' ',
+  imageFit: 'fill',
 };
 /* eslint-enable react/default-props-match-prop-types */
 
 class Image extends React.Component {
   constructor(props) {
     super(props);
 
+    this.myRef = React.createRef();
```
Comment (nramamurth) [inline packages/terra-image/src/Image.jsx]: This could have a better name, maybe?

File: packages/terra-image/src/Image.jsx [hunk 5]
```diff
@@ -117,23 +133,28 @@ class Image extends React.Component {
...
       'image',
+      imageFit,
       variant,
       customProps.className,
       { fluid: isFluid },
     ]);
     delete customProps.className;
-
+    if (!this.state.isLoading) {
+      const img = this.myRef.current;
```
Comment (nramamurth) [inline packages/terra-image/src/Image.jsx]: I'd do directly do - 

```
objectFitImages(this.imageRef.current);
```

Comment (pranav300) [reply packages/terra-image/src/Image.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/0127dec949032aa00c86bb77ebdeb92d3d695832.

File: packages/terra-image/src/Image.jsx (hunk 4 above)
Comment (pranav300) [reply packages/terra-image/src/Image.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/0127dec949032aa00c86bb77ebdeb92d3d695832.

File: packages/terra-avatar/src/common/AvatarUtils.jsx (hunk 3 above)
Comment (pranav300) [reply packages/terra-avatar/src/common/AvatarUtils.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/0127dec949032aa00c86bb77ebdeb92d3d695832.

File: packages/terra-image/src/Image.module.scss (hunk 2 above)
Comment (pranav300) [reply packages/terra-image/src/Image.module.scss]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/0127dec949032aa00c86bb77ebdeb92d3d695832.

File: packages/terra-image/src/Image.jsx [hunk 6]
```diff
@@ -49,20 +58,27 @@ const propTypes = {
    * Function to be executed when the image load errors.
    */
   onError: PropTypes.func,
+  /**
+  * Sets the `object-fit` style of the image from the following values; `cover`, `contain`, `fill`, `scale-down`, `none`.
+  */
+  imageFit: PropTypes.oneOf(['cover', 'contain', 'fill', 'scale-down', 'none']),
```
Comment (yuderekyu) [inline packages/terra-image/src/Image.jsx]: because we define these as an enumeration above, want to use object.values?
```suggestion
  imageFit: PropTypes.oneOf(Object.values(ImageFitTypes)),
```

File: packages/terra-avatar/src/common/AvatarUtils.jsx [hunk 7]
```diff
@@ -71,7 +71,8 @@ const generateImagePlaceholder = (alt, isAriaHidden, variant) => {
  */
 const generateImage = (image, alt, isAriaHidden, variant, handleFallback) => {
   const icon = generateImagePlaceholder(alt, isAriaHidden, variant);
-  return <TerraImage className={cx('image')} src={image} placeholder={icon} alt={alt} onError={handleFallback} />;
+  const imageFit = (variant === 'user') ? ImageFitTypes.COVER : undefined;
```
Comment (yuderekyu) [inline packages/terra-avatar/src/common/AvatarUtils.jsx]: The facility variant accepts an image prop too. Also, lets use the defined variants since they exist:

```suggestion
  const imageFit = (variant === AVATAR_VARIANTS.user || variant === AVATAR_VARIANTS.facility) ? ImageFitTypes.COVER : undefined;
```

Comment (pranav300) [reply packages/terra-avatar/src/common/AvatarUtils.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/700f775f5918c4a5babb6323f168d75b49205ec8.

Comment (pranav300) [reply packages/terra-avatar/src/common/AvatarUtils.jsx]: @bjankord Should we add test cases for non-square image for facility variant as well?

File: packages/terra-demographics-banner/CHANGELOG.md [hunk 8]
```diff
@@ -3,6 +3,9 @@ Changelog
 
 Unreleased
 ----------
+### Updated
```
Comment (avinashg1994) [inline packages/terra-demographics-banner/CHANGELOG.md]: nit: We should keep the CHANGELOG.md sections in alphabetical order, i.e. "Updated" section should come after "Changed".

Comment (pranav300) [reply packages/terra-demographics-banner/CHANGELOG.md]: I think changelog should be updated in the order in which changes were made.

File: packages/terra-image/src/Image.jsx [hunk 9]
```diff
@@ -49,20 +58,27 @@ const propTypes = {
    * Function to be executed when the image load errors.
    */
   onError: PropTypes.func,
+  /**
+  * Sets the `object-fit` style of the image from the following values; `cover`, `contain`, `fill`, `scale-down`, `none`.
+  */
+  imageFit: PropTypes.oneOf(Object.values(ImageFitTypes)),
```
Comment (mjhenkes) [inline packages/terra-image/src/Image.jsx]: Rename to fit? imageFit seems redundant.

File: packages/terra-image/src/Image.jsx [hunk 10]
```diff
@@ -12,6 +13,14 @@ const ImageVariant = {
   THUMBNAIL: 'thumbnail',
 };
 
+const ImageFitTypes = {
```
Comment (mjhenkes) [inline packages/terra-image/src/Image.jsx]: FitTypes?

File: packages/terra-image/src/Image.jsx [hunk 11]
```diff
@@ -49,20 +58,27 @@ const propTypes = {
    * Function to be executed when the image load errors.
    */
   onError: PropTypes.func,
+  /**
+  * Sets the `object-fit` style of the image from the following values; `cover`, `contain`, `fill`, `scale-down`, `none`.
```
Comment (mjhenkes) [inline packages/terra-image/src/Image.jsx]: Add some examples using fit to the documentation.

File: packages/terra-avatar/src/common/AvatarUtils.jsx [hunk 12]
```diff
@@ -71,7 +71,8 @@ const generateImagePlaceholder = (alt, isAriaHidden, variant) => {
  */
 const generateImage = (image, alt, isAriaHidden, variant, handleFallback) => {
   const icon = generateImagePlaceholder(alt, isAriaHidden, variant);
-  return <TerraImage className={cx('image')} src={image} placeholder={icon} alt={alt} onError={handleFallback} />;
+  const imageFit = (variant === AVATAR_VARIANTS.USER || variant === AVATAR_VARIANTS.FACILITY) ? ImageFitTypes.COVER : undefined;
```
Comment (mjhenkes) [inline packages/terra-avatar/src/common/AvatarUtils.jsx]: Is there harm in setting the fit to 'cover' for the shared user variant as well, just to skip the check?

If there is a problem, how do we address it for the default, no consumer supplied image, avatar and facility variants?

File: packages/terra-image/src/Image.jsx (hunk 9 above)
Comment (pranav300) [reply packages/terra-image/src/Image.jsx]: Fixed it here https://github.com/cerner/terra-core/pull/2494/commits/36742f3d9719b0fe00bbada1fcb416a338e7253e.

File: packages/terra-image/src/Image.jsx (hunk 10 above)
Comment (pranav300) [reply packages/terra-image/src/Image.jsx]: Fixed it here https://github.com/cerner/terra-core/pull/2494/commits/36742f3d9719b0fe00bbada1fcb416a338e7253e.

File: packages/terra-image/src/Image.jsx (hunk 11 above)
Comment (pranav300) [reply packages/terra-image/src/Image.jsx]: Fixed it here https://github.com/cerner/terra-core/pull/2494/commits/36742f3d9719b0fe00bbada1fcb416a338e7253e.

File: packages/terra-avatar/src/common/AvatarUtils.jsx (hunk 12 above)
Comment (pranav300) [reply packages/terra-avatar/src/common/AvatarUtils.jsx]: Fixed it here https://github.com/cerner/terra-core/pull/2494/commits/36742f3d9719b0fe00bbada1fcb416a338e7253e.

File: packages/terra-image/src/Image.jsx [hunk 13]
```diff
@@ -61,15 +61,15 @@ const propTypes = {
   /**
   * Sets the `object-fit` style of the image from the following values; `cover`, `contain`, `fill`, `scale-down`, `none`.
   */
-  imageFit: PropTypes.oneOf(Object.values(ImageFitTypes)),
+  Fit: PropTypes.oneOf(Object.values(FitTypes)),
```
Comment (mjhenkes) [inline packages/terra-image/src/Image.jsx]: lowercase 'fit' please  :grimacing:

File: packages/terra-avatar/src/common/AvatarUtils.jsx [hunk 14]
```diff
@@ -71,7 +71,7 @@ const generateImagePlaceholder = (alt, isAriaHidden, variant) => {
  */
 const generateImage = (image, alt, isAriaHidden, variant, handleFallback) => {
   const icon = generateImagePlaceholder(alt, isAriaHidden, variant);
-  return <TerraImage className={cx('image')} src={image} placeholder={icon} alt={alt} onError={handleFallback} />;
+  return <TerraImage className={cx('image')} src={image} placeholder={icon} alt={alt} onError={handleFallback} Fit={FitTypes.COVER} />;
```
Comment (mjhenkes) [inline packages/terra-avatar/src/common/AvatarUtils.jsx]: You could drop the additional import and just use the string.

```suggestion
  return <TerraImage className={cx('image')} src={image} placeholder={icon} alt={alt} onError={handleFallback} Fit={'cover'} />;
```

File: packages/terra-image/src/terra-dev-site/doc/example/ImageDefault.jsx [hunk 15]
```diff
@@ -27,6 +33,16 @@ const ImageDefault = () => (
...
+    <br />
+    <h2>Fit</h2>
+    <h3>Cover</h3>
+    <div className={cx('image-container')}>
+      <Image className={cx('image')} src={coverImage170X251} alt="invalid src attribute image" Fit="cover" variant="circle" />
+    </div>
+    <h3>None</h3>
+    <div className={cx('image-container')}>
+      <Image className={cx('image')} src={coverImage170X251} alt="invalid src attribute image" Fit="none" />
+    </div>
```
Comment (mjhenkes) [inline packages/terra-image/src/terra-dev-site/doc/example/ImageDefault.jsx]: I'm not sure these examples are really illustrating the use of this prop, it looks more like a difference between circle and square.

Take a look at mdn for inspiration 
https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit

File: packages/terra-avatar/src/common/AvatarUtils.jsx (hunk 14 above)
Comment (pranav300) [reply packages/terra-avatar/src/common/AvatarUtils.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/be42fad09a1ef9b1e6caf2b5042500e0e494d4db.

File: packages/terra-image/src/terra-dev-site/doc/example/ImageDefault.jsx (hunk 15 above)
Comment (pranav300) [reply packages/terra-image/src/terra-dev-site/doc/example/ImageDefault.jsx]: fixed here https://github.com/cerner/terra-core/pull/2494/commits/be42fad09a1ef9b1e6caf2b5042500e0e494d4db.

File: packages/terra-image/src/Image.module.scss [hunk 16]
```diff
@@ -36,4 +36,32 @@
...
+  /**
+   * Polyfill 'object-fit-images' for IE requires to use object-fit inside font-family.
+   */
+  .contain {
+    font-family: var(--terra-image-font-family, 'Helvetica Neue', Helvetica, Arial, sans-serif), 'object-fit: contain;';
```
Comment (mjhenkes) [inline packages/terra-image/src/Image.module.scss]: Does font family need to be themeable here?

Comment (pranav300) [reply packages/terra-image/src/Image.module.scss]: No, the font family does not need to be themeable. I have changed it here https://github.com/cerner/terra-core/pull/2494/commits/0ac279fc363761cc6d45b702fc9706bef1be5ac1.

File: packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx [hunk 17]
```diff
@@ -15,11 +15,11 @@ const ImageFluid = () => (
     </div>
     <p>Image width is larger than the container. Image height is smaller than the container.</p>
     <div id="height-smaller" className={cx('image-conatiner-reduced-width')}>
-      <Image src={placeholder150x150} alt="rounded" variant="rounded" isFluid />
+      <Image src={placeholder150x150} alt="rounded" variant="rounded" imageFit="cover" isFluid />
```
Comment (neilpfeiffer) [inline packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx]: @pranav300 prop typo:  imageFit --> fit="cover" (line 18 and line 22) so not actually applying in this test

File: packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx (hunk 1 above)
Comment (neilpfeiffer) [reply packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx]: wouldn't hurt to add a new test with each of the 'fit' values

File: packages/terra-profile-image/tests/jest/__snapshots__/ProfileImage.test.jsx.snap [hunk 18]
```diff
@@ -4,6 +4,7 @@ exports[`should render a Terra image with a placeholder prop 1`] = `
 <div>
   <Image
     alt="A profile image"
+    fit="fill"
```
Comment (neilpfeiffer) [inline packages/terra-profile-image/tests/jest/__snapshots__/ProfileImage.test.jsx.snap]: @pranav300 I'd expect this to always be `fit="cover"` like avatar (profile image was an older component that was used before avatar but in the same way), so lets add to it to `/ProfileImage.jsx` and update snapshot again.

**Comment by neilpfeiffer** [comment]
Tested: PR #2494
Resolves: Issue #2259
 
**Reviewed:**
- updates Terra-Image to allow for object-fit styling.
- updates Terra-Avatar and Terra-Profile-Image to always set `fit` to cover, as expected to best handle non-square images.

\
**Fix before release:**
- https://github.com/cerner/terra-core/pull/2494#pullrequestreview-270750706 and https://github.com/cerner/terra-core/pull/2494#pullrequestreview-270754890
 
 \
\
Marking as Verified by @neilpfeiffer, ready for release after adding fixes.
 
**Pass** |
-----------|
UX Review|

File: packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx (hunk 17 above)
Comment (pranav300) [reply packages/terra-image/src/terra-dev-site/test/image/ImageFluid.test.jsx]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/18dd5b93d46d856d085bf005ebcac7b4ac2b3685

File: packages/terra-profile-image/tests/jest/__snapshots__/ProfileImage.test.jsx.snap (hunk 18 above)
Comment (pranav300) [reply packages/terra-profile-image/tests/jest/__snapshots__/ProfileImage.test.jsx.snap]: Fixed here https://github.com/cerner/terra-core/pull/2494/commits/18dd5b93d46d856d085bf005ebcac7b4ac2b3685
